### PR TITLE
Fix #22932 - ensure correct behaviour when deleting segments used by spanners, such that undo/redo works without crashing

### LINKED ISSUE #22932: Rit ending inside measure - Clear measure+Undo+Redo and repeat causes crash
### Issue type

Crash or freeze

### Bug description

This crash is in master, not 4.3.1

### Steps to reproduce

1. Make a score with a rit (or other tempo line) ending on beat three of a measure
2. Select measure with end of rit
3. Clear, undo, redo
4. Click away and then repeat step 2 & 3
5. Crash

### Screenshots/Screen recordings

https://github.com/musescore/MuseScore/assets/69917893/afa767e8-ef5e-4bc1-a6a8-36761968232f

### MuseScore Version

OS: Windows 11 Version 2009 or later, Arch.: x86_64, MuseScore Studio version (64-bit): 4.4.0-241420305, revision: github-musescore-musescore-449a64e

### Regression

Yes, between 4.3.1 and 4.4 (master)

### Operating system

All

### Additional context

[Crash.zip](https://github.com/musescore/MuseScore/files/15395086/Crash.zip)

### PR DESCRIPTION
**Pull request by wizofaus**
Resolves: #22932

- [x] I signed the [CLA](https://musescore.org/en/cla)
- [x] The title of the PR describes the problem it addresses
- [x] Each commit's message describes its purpose and effects, and references the issue it resolves
- [x] If changes are extensive, there is a sequence of easily reviewable commits
- [x] The code in the PR follows [the coding rules](https://github.com/musescore/MuseScore/wiki/CodeGuidelines)
- [x] There are no unnecessary changes
- [x] The code compiles and runs on my machine, preferably after each commit individually
- [ ] I created a unit test or vtest to verify the changes I made (if applicable)

### CODE REVIEW

File: src/engraving/dom/edit.cpp [hunk 1]
```diff
@@ -6519,7 +6526,13 @@ void Score::undoRemoveElement(EngravingItem* element, bool removeLinked)
             if (s->header() || s->trailer()) {        // probably more segment types (system header)
                 s->setEnabled(false);
             } else {
+                auto needAnchor = isSpannerStartEnd(s->tick(), element->track());
                 doUndoRemoveElement(s);
+                if (needAnchor) {
+                    TimeTickAnchor* anchor = EditTimeTickAnchors::createTimeTickAnchor(s->measure(), s->tick() - s->measure()->tick(), track2staff(element->track()));
+                    doUndoAddElement(anchor->parentItem());
+                    doUndoAddElement(anchor);
```
Comment (wizofaus) [inline src/engraving/dom/edit.cpp:6534]: code duplication here, if it's agreed that this is the correct fix, will clean up

File: src/engraving/dom/score.cpp [hunk 2]
```diff
@@ -4395,7 +4395,7 @@ void Score::removeSpanner(Spanner* s)
 bool Score::isSpannerStartEnd(const Fraction& tick, track_idx_t track) const
 {
     for (auto i : m_spanner.map()) {
-        if (i.second->track() != track) {
+        if (i.second->track() != track && !i.second->isGradualTempoChange()) {
```
Comment (wizofaus) [inline src/engraving/dom/score.cpp:4398]: My expectation was to check the "isSystem()" property, but this is *not* true for gradual tempo changes.  But gradual tempo changes can end up in a different staff to the segments that are selected when you add them - they appear to be only type of element for which this is true.

**Comment by wizofaus** [comment]
Came across quite a few concerning things looking at this, I'm really not sure we should ever have code that creates segments as part of doing layout (when there's no opportunity to safely update the undo stack), but that's what was happening - this code essentially prevents that from triggering by ensuring that there are always segments at any needed locations, and ensures they're properly tracked via the undo stack.

**Comment by mike-spa** [comment]
Sorry, I can't merge this solution. Anchors are meant to work like an invisible ruler that the system creates for itself "on the spot" whenever they are needed. I think they really shouldn't be in the undo stack because they are not something that you actually _do_ or undo. So I think the correct solution is the opposite actually, i.e. not adding them to the stack, but ensuring that they never get there. I need to look into this myself, cause anchors are a very new piece of the system and all these interactions need to be managed carefully

**Comment by wizofaus** [comment]
But that's what it *is* doing - it's creating anchors because it knows they will be needed once the chordrest segments are gone.

In actual fact, I'm looking at a similar issue now where MMRest objects are created *during* layout - but it does in fact do them as part of the undo stack.  I can't say it's the way I'd prefer to see it done, but it seems in fact quite a lot of objects can get created during the layout that happens immediately after a score edit, and as long as those changes go on the undo stack, it seems to do the right thing.  So the alternative change is to ensure that the segments that were created automatically as part of layout are done in a way that is safe in respect to the undo stack. Note the basic sequence I was seeing was:

a) as part of layout, segment is added to measure (but nothing updated in undo stack)
b) on undo of clear, that segment is not removed, but it attempts to ANOTHER segment at the same place, however this does nothing because it has logic to ensure duplicate segments are not created
c) on redo, it attempts to remove the segment that the previous undo had tried to add, but this segment doesn't actually exist in the measure

**Comment by mike-spa** [comment]
You may well be right, but I still prefer to look into this issue myself, cause it's a very recent system that I've designed and I consider it still a work in progress, meaning that any decision we make now is very consequential, and therefore quite delicate. I'll come back to you.

In the meantime, we'd like to encourage you to look at the community-tagged issues, where any help you can give is surely appreciated!

**Pull request closed without merging**